sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -1034,16 +1034,35 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
 };
 
 /// Called in device code. This function iterates through the index space
-/// \p Range using stride equal to the global range specified in \p NdId,
+/// by assigning contiguous chunks to each work-group, then iterating
+/// through each chunk using a stride equal to the work-group's local range,
 /// which gives much better performance than using stride equal to 1.
 /// For each of the index the given \p F function/functor is called and
 /// the reduction value hold in \p Reducer is accumulated in those calls.
 template <typename KernelFunc, int Dims, typename ReducerT>
-void reductionLoop(const range<Dims> &Range, ReducerT &Reducer,
+void reductionLoop(const range<Dims> &Range, const size_t PerGroup,
+                   const size_t Remainder, ReducerT &Reducer,
                    const nd_item<1> &NdId, KernelFunc &F) {
-  size_t Start = NdId.get_global_id(0);
-  size_t End = Range.size();
-  size_t Stride = NdId.get_global_range(0);
+
+  // Divide into contiguous chunks and assign each chunk to a Group
+  // Rely on precomputed PerGroup and Remainder to avoid repeating modulo
+  auto Group = NdId.get_group();
+  size_t GroupId = Group.get_group_linear_id();
+  size_t NumGroups = Group.get_group_linear_range();
+  size_t GroupStart = GroupId * PerGroup;
+  size_t GroupEnd = GroupStart + PerGroup;
+  bool LastGroup = (GroupId == NumGroups - 1);
+  if (LastGroup) {
+    GroupEnd += Remainder;
+  }
+  if (GroupEnd > Range.size()) {
+    GroupEnd = Range.size();
+  }
```

Comment:
Can't these lines be replaced with simpler code? It would also help to eliminate the 'Reminder' parameter.
```suggestion
  bool LastGroup = (GroupId == NumGroups - 1);
  size_t GroupEnd = LastGroup ? (GroupStart + PerGroup : Range.size());
```

Reply:
Thanks; that is simpler!  I've made the change in b3452366c7fd90bde32e742c540f75ea56d52d49.

I kept the `Remainder` parameter because I think it will be useful for anybody who wants to experiment with other division strategies or specialize for different devices.

Reply:
Hmm, _potentially_ may be useful to somebody. Seems like this would be better if it removed and not confuse anyone.

Reply:
Ok, I'll remove it.  But leaving a note here for posterity...

Passing in the remainder would allow the division to be adjusted like so:
```c++
if (GroupId < Remainder) {
  GroupStart += GroupId;
  GroupEnd += GroupId;
 }
else {
  GroupStart += Remainder;
  GroupEnd += Remainder;
}
```

This would result in more even distribution of iterations to work-groups and reduce load imbalance.  But it's unclear to me whether there are some devices that wouldn't like this, or if we should really be making decisions about how to handle the remainder based on sub-group size.

Reply:
Removed in 782b29caff92bf0f5ba32f1fa93f9f52fa2d2e81, and I left a TODO.